[browser][MT] WebSocket thread affinity & server initiated close

pavelsavara opened the pull request:
- `BrowserWebSocket` now uses automatic dispatch to correct thread via `JSSynchronizationContext` & `JSProxyContext` & JS object affinity
- `onClosed` event doesn't call back to C#
   - simplifies locking for MT
   - doesn't allocate Delegate/Function proxy
- fix `CloseReceived` and `CloseSent` states
- enabled `CloseAsync_ServerInitiatedClose_Success`
- new test `CloseOutputAsync_ServerInitiated_CanReceive`
- `System.Net.WebSockets.Client.Tests` now also run xunit on thread pool in MT
- cross-thread WS tests in `WebWorkerTest`

Fixes https://github.com/dotnet/runtime/issues/65796
Fixes https://github.com/dotnet/runtime/issues/96173

Contributes to https://github.com/dotnet/runtime/issues/96359 
With partial/loose backport for Net8 https://github.com/dotnet/runtime/pull/97002

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
rebase

pavelsavara committed:
Merge branch 'main' into browser_ws_dispatch

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-96618-merge-37ec7d9659c447a6b9/WasmTestOnBrowser-System.Runtime.InteropServices.JavaScript.Tests/1/console.621c9e3d.log?helixlogtype=result)
```
[16:58:44] info: [2024-01-11T16:58:44.253Z] [FAIL] System.Runtime.InteropServices.JavaScript.Tests.WebWorkerTest.JSSynchronizationContext_Send_Post_Items_Cancellation
[16:58:44] info: System.OperationCanceledException : The operation was canceled.
[16:58:44] info:    at System.Runtime.InteropServices.JavaScript.JSSynchronizationContext.Post(SendOrPostCallback , Object )
[16:58:44] info:    at System.Runtime.InteropServices.JavaScript.Tests.WebWorkerTest.<>c__DisplayClass6_0.<JSSynchronizationContext_Send_Post_Items_Cancellation>b__2()
[16:58:44] info:    at System.Threading.Tasks.Task`1[[System.Threading.Tasks.Task, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
[16:58:44] info:    at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj)
[16:58:44] info:    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread , ExecutionContext , ContextCallback , Object )
[16:58:44] info: --- End of stack trace from previous location ---
[16:58:44] info:    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread , ExecutionContext , ContextCallback , Object )
[16:58:44] info:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& , Thread )
[16:58:44] info: --- End of stack trace from previous location ---
[16:58:44] info:    at System.Runtime.InteropServices.JavaScript.Tests.WebWorkerTest.JSSynchronizationContext_Send_Post_Items_Cancellation()
[16:58:44] info: --- End of stack trace from previous location ---
```
and
```
[16:58:47] info: [2024-01-11T16:58:47.198Z] [FAIL] System.Runtime.InteropServices.JavaScript.Tests.WebWorkerTest.WebSocketClient_ResponseCloseInDifferentThread(executor1: JSWebWorker, executor2: JSWebWorker)
[16:58:47] info: System.OperationCanceledException : The operation was canceled.
[16:58:47] info:    at System.Runtime.InteropServices.JavaScript.Tests.WebWorkerTest.<ActionsInDifferentThreads>d__20`1[[System.Net.WebSockets.ClientWebSocket, System.Net.WebSockets.Client, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
[16:58:47] info: --- End of stack trace from previous location ---
```

[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-96618-merge-021e25678133440888/WasmTestOnBrowser-System.Net.WebSockets.Client.Tests/1/console.ffbd46da.log?helixlogtype=result)
```
[16:45:55] info: [2024-01-11T16:45:55.982Z] [STRT] System.Net.WebSockets.Client.Tests.InvokerConnectTest.ConnectAsync_PassNoSubProtocol_ServerRequires_ThrowsWebSocketException(server: ws://127.0.0.1:43735/WebSocket/EchoWebSocket.ashx)
[16:45:55] warn: MONO_WASM [0xdbdfc-main]: WebSocket error {}
[16:45:55] fail: MONO_WASM: {"status":1,"silent":true}
[16:45:56] fail: [out of order message from the browser]: http://127.0.0.1:43735/_framework/dotnet.runtime.js 2 WebSocket connection to 'ws://127.0.0.1:43735/WebSocket/EchoWebSocket.ashx?subprotocol=CustomProtocol' failed: Error during WebSocket handshake: Response must not include 'Sec-WebSocket-Protocol' header if not present in request: CustomProtocol
[16:45:56] info: WASM EXIT 1
[16:45:56] info: Waiting to flush log messages with a timeout of 120 secs ..
[16:45:56] fail: Application has finished with exit code TESTS_FAILED but 0 was expected
XHarness exit code: 71 (GENERAL_FAILURE)
```

pavelsavara committed:
Merge branch 'main' into browser_ws_dispatch

kg commented on src/libraries/System.Net.WebSockets.Client/tests/CloseTest.cs:
```diff
@@ -33,12 +33,12 @@ public class CloseTest : ClientWebSocketTestBase
         public async Task CloseAsync_ServerInitiatedClose_Success(Uri server, bool useCloseOutputAsync)
         {
-            const string closeWebSocketMetaCommand = ".close";
+            const string closeWebSocketMetaCommand = ".shutdown";
```
should rename this variable since it's no longer close

kg commented on src/libraries/System.Net.WebSockets.Client/tests/CloseTest.cs:
```diff
@@ -69,9 +69,10 @@ public async Task CloseAsync_ServerInitiatedClose_Success(Uri server, bool useCl
 
                 // Send back close message to acknowledge server-initiated close.
                 _output.WriteLine("Close starting.");
+                var closeStatus = PlatformDetection.IsNotBrowser ? WebSocketCloseStatus.InvalidMessageType : (WebSocketCloseStatus)3210;
```
is the number 3210 important?

pavelsavara replied:
It needs to be in range that the browser can send. See https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#code

pavelsavara committed:
Merge branch 'main' into browser_ws_dispatch

kg commented on src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:
```diff
@@ -249,12 +249,12 @@ public async Task SendAsync_MultipleOutstandingSendOperations_Throws(Uri server)
             using (ClientWebSocket cws = await GetConnectedWebSocket(server, TimeOutMilliseconds, _output))
             {
-                var cts = new CancellationTokenSource(TimeOutMilliseconds);
+                var cts = new CancellationTokenSource(PlatformDetection.LocalEchoServerIsNotAvailable ? TimeOutMilliseconds : 200);
```
Why 200?

pavelsavara replied:
with local web server 200ms should be enough for it to work. When it doesn't work, waiting 5 minutes for timeout was too much.

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -33,122 +34,126 @@ public override WebSocketState State
-                return FastState = _innerWebSocket!.SynchronizationContext.Send(static (BrowserWebSocket self) =>
-                {
-                    lock (self._thisLock)
+                    var st = GetReadyState(_innerWebSocket!);
```
this call is synchronously remoted to the thread that owns the socket, right?

pavelsavara replied:
right, synchronously in sense, that the sender waits for the WS thread to read and respond. There could be other un-ordered actions from other threads, sending into same WS thread.

I think that I could afford to call it from locked section, because it doesn't make call any other action inside, which could lead to deadlock.

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -199,26 +191,13 @@ public override Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buf
             cancellationToken.ThrowIfCancellationRequested();
-            ThrowIfDisposed();
 
             WebSocketValidate.ValidateCloseStatus(closeStatus, statusDescription);
```
i feel like all this state validation should occur inside the lock

pavelsavara replied:
`ThrowIfDisposed` contains lock inside

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -189,7 +181,7 @@ public override Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buf
                 return Task.FromException<WebSocketReceiveResult>(new OperationCanceledException(cancellationToken));
             }
-            ThrowIfDisposed();
+
```
shouldn't this whole method occur inside the lock?

pavelsavara replied:
refactored

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-                    }
+                    return Task.CompletedTask;
+                }
+                if (state == WebSocketState.None || state == WebSocketState.Closed)
```
Two cases for Closed here

pavelsavara replied:
good catch, thanks!

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -33,122 +34,126 @@ public override WebSocketState State
-#endif
+                        if (_closeReceived && _closeSent)
+                        {
+                            st = WebSocketState.Closed;
```
Why do we have this condition and the last one? The `st` is already set to `Closed` or `CloseSent` when we reach this block

pavelsavara replied:
The JS side state doesn't have separate `closeSent` and `closeReceived` states. That's why we have these bools to make sense of it.

maraf replied:
Oh, I see. The state can be `CloseSend` and we override it to `Closed` and similar on line 55, state can be `Closed` and we override it to `CloseSent`. That's how it is supposed to work, right?

maraf replied:
Oh, I see. The state can be `CloseSend` and we override it to `Closed` and similar on line 55, state can be `Closed` and we override it to `CloseSent`. That's how it is supposed to work, right?

pavelsavara replied:
right

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -227,37 +206,24 @@ public override Task CloseOutputAsync(WebSocketCloseStatus closeStatus, string?
-#else
-            return CloseAsyncCore(closeStatus, statusDescription, false, cancellationToken);
-#endif
+                    throw new WebSocketException(WebSocketError.InvalidState, SR.Format(SR.net_WebSockets_InvalidState, state, "Connecting, Open, CloseSent, Aborted"));
```
CloseAsync while Closed returns CompletedTask now, shouldn't CloseOutputAsync while Closed also return CompletedTask?

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-#endif
+                state = State;
+            } // lock
+            AbortCore(this, state);
```
is it intentional that we no longer abort within the lock?

pavelsavara replied:
refactored

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-        {
-#if FEATURE_WASM_THREADS
-            lock (_thisLock)
+            if (state < WebSocketState.Closed && state != WebSocketState.None)
```
i would prefer to see all this logic happen inside the lock but i'm not sure if that's feasible. i do think this code will misbehave if two threads race on it

pavelsavara replied:
refactored, please have look again

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -33,122 +34,126 @@ public override WebSocketState State
+                ThrowIfDisposed();
+                lock (_lockObject)
+                {
+                    if (_innerWebSocket == null) throw new InvalidOperationException(SR.net_WebSockets_NotConnected);
```
The `BrowserInterop.GetProtocol` is returning `null` when web socket null or disposed, shouldn't we use the same behavior here as well?

pavelsavara replied:
this is pre-existing behavior, I'll not change it as part of this PR

maraf replied:
I don't understand. You are changing it, don't you?

pavelsavara replied:
no, I'm not. This is the old one

![image](https://github.com/dotnet/runtime/assets/271576/3ffec89d-4047-4e48-908d-9bdf952e4838)

kg commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-            if (!self._disposed && currentState != WebSocketState.Closed)
+            lock (self._lockObject)
             {
                 self.FastState = WebSocketState.Aborted;
```
i think it's incorrect that we are able to change the state or aborted flag if we're already disposed. should it be possible to change internal state on a disposed socket?

pavelsavara replied:
I updated the code. We have now 2 abort scenarios
- called directly, which will do blocking abort
- called from `Dispose`, which will schedule the abort onto the owner thread. In this case `_aborted` will change after `_disposed`

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -33,122 +34,126 @@ public override WebSocketState State
             // fast check of previous _state instead of GetReadyState(), the readyState would be validated on JS side
-            if (FastState != WebSocketState.Open)
+            var fastState = FastState;
+            if (fastState != WebSocketState.Open && fastState != WebSocketState.CloseReceived)
```
What will happen next when the state is `CloseReceived`?

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-                Task promise;
-                lock (_thisLock)
+                state = State;
+                if (state == WebSocketState.Closed)
```
Previously `Closed` state was throwing an expcetion, now it's ok, is it intended?

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -471,160 +375,247 @@ private async Task ConnectAsyncCore(CancellationToken cancellationToken)
+                        _cancelled = true;
+                        throw new OperationCanceledException(cancellationToken);
+                    }
+                    if (state != WebSocketState.Closed && ex.Message == "Error: OperationCanceledException")
```
Can we do anything better than checking exception message (storing a flow on the exception or anything?

pavelsavara replied:
this is rather fallback in case `cancellationToken.IsCancellationRequested` didn't catch it. It's not worth the trouble to complicate it, we own the message so it would not change on it's own.

maraf replied:
I don't thing it will be obvious when we do localization for example, but it's your call

pavelsavara commented on src/mono/browser/runtime/web-socket.ts:
```diff
@@ -147,18 +142,23 @@ export function ws_wasm_receive(ws: WebSocketExtension, buffer_ptr: VoidPtr, buf
     const receive_promise_queue = ws[wasm_ws_pending_receive_promise_queue];
 
-    const readyState = ws.readyState;
-    if (readyState != WebSocket.OPEN && readyState != WebSocket.CLOSING) {
```
because it was wrong. it must be possible to read pending messages even if the sending side close their side of the WS. We have user issue for that.

pavelsavara committed:
feedback

- calling methods from inside of the lock, but awaiting outside
- made Dispose->Abort async to target thread
- optimized CancellationHelper
- added wasm_ws_close_received and wasm_ws_close_received on JS side
- added extra state validation on JS side
- removed mono_wasm_web_socket_close_warning

pavelsavara commented:
/azp run runtime-wasm

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserInterop.cs:
```diff
@@ -16,6 +16,24 @@ internal static partial class BrowserInterop
 
+        public static WebSocketCloseStatus? GetCloseStatus(JSObject? webSocket)
+        {
+            if (webSocket == null || webSocket.IsDisposed) return null;
```
```suggestion
            if (webSocket == null || webSocket.IsDisposed)
                return null;
```

maraf commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserInterop.cs:
```diff
@@ -16,6 +16,24 @@ internal static partial class BrowserInterop
+
+        public static string? GetCloseStatusDescription(JSObject? webSocket)
+        {
+            if (webSocket == null || webSocket.IsDisposed) return null;
```
```suggestion
            if (webSocket == null || webSocket.IsDisposed)
                return null;
```

pavelsavara committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserInterop.cs

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserInterop.cs

Co-authored-by: Marek Fišera <mara@neptuo.com>

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
whitespace

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm
